Working prototype with TOML configuration and CLI interface

sunberry00 opened the pull request.

akomii committed:
Revert "First Pull Request" (#3)

Commit:
Change structure of project

Commit:
Correct file extension from .xml to .cda

Commit:
Add config.toml file with path variables

Commit:
Add --rows parameter to command line to parse number of patients should be generated

Commit:
Small adjustments

Commit:
Add given and family names for patient and author

Commit:
Add exception for link if file not exists

Commit:
Small changes in formatting

Commit:
Change README.md

Commit:
Replace if-else cascade with map

Commit:
Add csv with all possible value sets

Commit:
Add header to .csv

Commit:
Add new parameter "column" which represent a column that should be used in .csv. Every link parameter should be followed by "link" parameter

Commit:
Implement Diagnostik

Commit:
Small formatting

Commit:
Add docstring and rename functions

Commit:
Add docstring and rename functions

Commit:
Fix typo

Commit:
Fix reading number of datasets from CLI

Commit:
Refactor generator.py, removed duplicate code and Add docstrings

Commit:
Add for city and corresponding organisation a table

Commit:
Add streets and postalcodes.

Commit:
Addd correct GCS Score

Commit:
Add correct pregnancy

akomii commented on resources/value_sets/cities.csv:
```diff
@@ -0,0 +1,78 @@
+city;klinik_name;postleitzahl
+Berlin;Charité – Universitätsmedizin Berlin;10117
+Hamburg;Universitätsklinikum Hamburg-Eppendorf (UKE);20246
```
For me it seems instead of cities, the focus is of clinics in this csv.

akomii commented on resources/value_sets/diagnostik.csv:
```diff
@@ -0,0 +1,16 @@
+displayName;diagnostik_id;diagnostik_code
+Chest X-ray;1.2.276.0.76.10.4050;30745-4
```
german file name..

Commit:
Add Associated Person for patients with family insurance

Commit:
Refactor csv_to_cda, raw_to_CDA. Add configuration for output folder

Commit:
Add new ToDos

Commit:
main_and_sent_test() implementiert

Commit:
main_and_sent_test() small adjustments

Commit:
calculate dependency: refactored code added tasks

Commit:
added new Variable: _diagnostik_id

Commit:
gcs Sum and pregnent refactored

Commit:
associated Person in Excel eingefügt nicht mehr in dependencies; added docs in dependencies

Commit:
utf8 fixed

Commit:
XSLT prüfen und Variablen überarbeiten bis keime

Commit:
XSLT prüfen und Variablen überarbeiten bis tetanus

Commit:
XSLT prüfen und Variablen überarbeiten fertig

Commit:
Add all Diagnostiken (Procedures) to XSLT

Commit:
Dignostik alle Felder in XSLT

Commit:
Dignostik implementiert und refactoring

Commit:
Dignostik implementiert und refactoring

Commit:
Nullflavors implementiert für alle die einen Nullflavor haben

Commit:
Add async method to send all cda at once

Commit:
Add icd10 codes

Commit:
Add NullFlavor support for Transportmittel

Commit:
Add NullFlavor support for CEDIS

Commit:
Fix Cedis and OriginalText BUG

Commit:
Add NullFlavors for Diagnose ICD10 and function with statistics

Commit:
Mehrere Diagnosen mit Qualifier added

Commit:
Diagnosen qualifier wird immer gesetzt bei code=unk originaltext=unk

Commit:
Refactoring

Commit:
Fix performance issue with adding new columns to DataFrame

Commit:
Formatting

Commit:
Add docstring

Commit:
Formatting

Commit:
Refactor calculate_timestamps to be column-wise

akomii commented on resources/value_sets/streets.csv:
```diff
@@ -0,0 +1,51 @@
+street_address_line
+Hauptstraße
+Dorfstraße
```
Maybe merge this file together with family_names and first_names to reduce file numbers?

akomii commented on src/calculate_dependencies.py:
```diff
@@ -1,36 +1,167 @@
+        pd.DataFrame: The updated DataFrame with calculated timestamp dependencies.
+    """
+
+    def add_minutes_to_timestamp(timestamp, minutes, format_in="%Y%m%d%H%M%S", format_out="%Y%m%d%H%M%S"):
```
Instead of making a nested method, a private method would be easier to read

akomii commented on src/main.py:
```diff
@@ -1,20 +1,72 @@
+
+
+def main():
+    global config, parser, args, rows
```
Why are these globals? Atleast config is not necessary, because you can access environment variables from everywhere as long as the process lives

akomii commented on src/main.py:
```diff
@@ -1,20 +1,72 @@
+def parse_command_line():
+    global parser, args, rows
+    parser = argparse.ArgumentParser(description='Process Excel to CDA.')
+    parser.add_argument('--rows', type=int, required=True, help='Number of rows to generate in the CSV file')
```
--rows is too abstract. Better would be "--n" or "--encounter"

akomii commented on src/main.py:
```diff
@@ -1,20 +1,72 @@
+    rows = args.rows
+
+
+def main():
```
When running the script I get following errors:

```
Traceback (most recent call last):
File "/home/akomi/PycharmProjects/cda-test-data-generator/src/main.py", line 72, in <module>
main()
File "/home/akomi/PycharmProjects/cda-test-data-generator/src/main.py", line 65, in main
csv_to_cda(csv_path, xslt_file)
File "/home/akomi/PycharmProjects/cda-test-data-generator/src/csv_to_cda.py", line 115, in csv_to_cda
output_dir = os.environ['output_dir']
File "/usr/lib/python3.10/os.py", line 680, in __getitem__
raise KeyError(key) from None
KeyError: 'output_dir'
Error in sys.excepthook:
Traceback (most recent call last):
File "/usr/lib/python3/dist-packages/apport_python_hook.py", line 72, in apport_excepthook
from apport.fileutils import likely_packaged, get_recent_crashes
File "/usr/lib/python3/dist-packages/apport/__init__.py", line 5, in <module>
from apport.report import Report
File "/usr/lib/python3/dist-packages/apport/report.py", line 32, in <module>
import apport.fileutils
File "/usr/lib/python3/dist-packages/apport/fileutils.py", line 28, in <module>
from apport.packaging_impl import impl as packaging
File "/usr/lib/python3/dist-packages/apport/packaging_impl.py", line 23, in <module>
import apt
ModuleNotFoundError: No module named 'apt'

Original exception was:
Traceback (most recent call last):
File "/home/akomi/PycharmProjects/cda-test-data-generator/src/main.py", line 72, in <module>
main()
File "/home/akomi/PycharmProjects/cda-test-data-generator/src/main.py", line 65, in main
csv_to_cda(csv_path, xslt_file)
File "/home/akomi/PycharmProjects/cda-test-data-generator/src/csv_to_cda.py", line 115, in csv_to_cda
output_dir = os.environ['output_dir']
File "/usr/lib/python3.10/os.py", line 680, in __getitem__
raise KeyError(key) from None
KeyError: 'output_dir'
```

sunberry00 replied:
I don't really know how can I reproduce this behaviour

akomii commented on src/generate_csv.py:
```diff
@@ -1,57 +1,119 @@
+    Extract variables with their default values, generation type, and parameters from an Excel input.
 
+    Args:
+        excel_input (pd.DataFrame): The input DataFrame read from the Excel file.
```
But a DataFrame is a DataFrame. It doesnt matter if the binary file was Excel or CSV... Rename the var and recheck the docs

akomii commented on src/main.py:
```diff
@@ -1,20 +1,72 @@
+    # Get environment variables
+    csv_path = os.environ['CSV_PATH']
+    excel_path = os.environ['EXCEL_PATH']
+    xslt_file = os.environ['XSLT_FILE']
```
All inline comments above are redundant. Remove them.

akomii commented on src/generate_csv.py:
```diff
@@ -1,57 +1,119 @@
     return variables_dict
 
 
 def generate_csv(excel_path: str, csv_path, num_datasets=1) -> None:
```
This method is too long. Seperate the concerns. A method should be only around 20 LOC

akomii replied:
Why is num_datasets optional? The main.py throws in error if --rows is empty

akomii commented on src/generate_csv.py:
```diff
@@ -1,57 +1,119 @@
+        if var_type in types:
+            # Generate data column
+            generator = GeneratorFactory.create_generator(GeneratorType(var_type), params).generate()
+            column_list = generate_column_list(generator, num_datasets, null_flavors, 0.5)
```
probability_missing should also be configurable

akomii commented on src/generator.py:
```diff
@@ -0,0 +1,230 @@
+    DATE = 'date'
+
+
+class AbstractGenerator(ABC):
```
What value do the docs in this class provide?

akomii commented on src/generator.py:
```diff
@@ -0,0 +1,230 @@
+    Generator for random UUID values.
+    """
+
+    def __init__(self, **kwargs):
```
redundant

akomii commented on src/generator.py:
```diff
@@ -0,0 +1,230 @@
+    Generator for random dates within a specified range.
+    """
+
+    def __init__(self, **kwargs):
```
Why are you using keyword-arguments? AbstractGenerator does not provide a constructor so you can just overwrite it like:

```python
    def __init__(self, start_date=None, end_date=None, date_format="yyyymmddhhmmss"):
        self.start_date = start_date if start_date else datetime(2000, 1, 1)
        self.end_date = end_date if end_date else datetime.today()
        self.date_format = date_format
```

akomii commented on src/generator.py:
```diff
@@ -0,0 +1,230 @@
+    Generator for random string values based on a value set or regex pattern.
+    """
+
+    def __init__(self, **kwargs):
```
Too much LOC in the constructor

akomii replied:
```python
    def __init__(self, value_set=None, regex=None, link=None, column=None):
        self.value_set = value_set
        self.regex = regex
        if link:
            self._load_value_set_from_csv(link, column)

    def _load_value_set_from_csv(self, link, column):
        if not column:
            raise ValueError("Column not specified in parameters")
        df = pd.read_csv(f"../resources/value_sets/{link}", delimiter=";", dtype=str, header=0)
        if column not in df.columns:
            raise ValueError(f"Column '{column}' not found in file")

        self.value_set = set(df[column])
```

akomii commented on src/generator.py:
```diff
@@ -0,0 +1,230 @@
+                yield exrex.getone(self.regex)
+        else:
+            while True:
+                yield ''.join(random.choice('ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789') for _ in range(10))
```
Its hard too know what exactly happens in each if-clause. Either add descriptive inline comments or add methods for better readability

akomii replied:
```python
    def __generate_from_value_set(self):
           while True:
            yield random.choice(tuple(self.value_set))

    def __generate_from_regex(self):
            while True:
            yield exrex.getone(self.regex)

    def __generate_random_alphanumeric():
           while True:
            yield ''.join(random.choice('ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789') for _ in range(10))

    def generate(self):
        if self.value_set:
            __generator = self._generate_from_value_set()
        elif self.regex:
            __generator = self._generate_from_regex()
        else:
            __generator = self._generate_random_alphanumeric()

        yield from generator
```

akomii commented on src/generator.py:
```diff
@@ -0,0 +1,230 @@
+        """
+        try:
+            generator_class = cls._generator_map[generator_type]
+        except:
```
specify the exception

akomii commented on src/calculate_dependencies.py:
```diff
@@ -1,36 +1,167 @@
 import datetime as dt
 
+from generator import GeneratorFactory, GeneratorType
+
```
Overall good, check again for redundancy of comments and separation of concerns.

akomii commented on src/parser.py:
```diff
@@ -1,3 +1,4 @@
+import os
 from datetime import datetime
 
```
Descriptive comments are missing. What exactly do they handle?

akomii reviewed (changes requested):
Good Job :thumbs_up Make sure to find the balance between extensive and redudant documentation. Also recheck parts of your business logic. Often its better for readability to move processing steps into separate methods

Commit:
Change name and order of columns in cities.csv to clinics.csv with more focus on clinics

Commit:
Change name to diagnostics.csv

Commit:
Merge various .csv files in one misc.csv (miscellaneous)

Commit:
Change nested method as a private

Commit:
Remove global variables

Commit:
Change rows to n

Commit:
Check if output/responses exists. If not, create

Commit:
Rename extract_concept_id_attributes Args and docs

Commit:
Remove redundant inline comments

Commit:
Remove redundant docs

Commit:
Remove redundant docs

Commit:
Change format parameter -> date_format and removed **kwargs in DateGenerator

Commit:
Remove "number" argument before variable were generated

Commit:
Refactor StringGenerator

Commit:
Refactor StringGenerator

Commit:
Change Exception to KeyError

Commit:
Refactor calculate_dependencies.py, remove redundant comments etc.

Commit:
Refactor calculate_dependencies.py, remove redundant comments etc.

Commit:
Add docs to parser.py

Commit:
Extract some logic to separate methods

Commit:
Make num_datasets not optional

Commit:
Add probability_missing parameter

Commit:
Correct requirements.txt

akomii commented on resources/value_sets/cedis.csv:
```diff
@@ -1,3 +1,4 @@
+cedis;displayName
```
snake-case is the standard in python

akomii commented on resources/value_sets/clinics.csv:
```diff
@@ -0,0 +1,78 @@
+klinik_name;city;postleitzahl
```
english or german? Choose one and stick with it (choose english)

JohannesKramer replied:
Die Variablen Namen stimmen mit den Concept-Ids aus Aktin überein (Die sind auf Deutsch). Bei Variablen die nicht in Aktin sind haben wir uns einen namen ausgedacht wir überarbeiten diese so, dass die einheitlich auf Deutsch sind.

akomii commented on resources/value_sets/misc.csv:
```diff
@@ -0,0 +1,101 @@
+first_name;family_name;street_address_line
```
just street_name would be enough

akomii replied:
Also misc.csv is not clear enough. Be precise

akomii commented on requirements.txt:
```diff
@@ -6,9 +6,11 @@ pip~=24.0
 numpy~=2.0.1
 setuptools~=72.1.0
-openpyxl~=3.2.0b1
+openpyxl
```
version is missing

akomii commented on src/main.py:
```diff
@@ -1,20 +1,67 @@
+
+
+def main():
+    config = toml.load('../config.toml')
```
There should be no relative paths for using this script... Provide the config.toml also as an input argument. If I run

```shell
python3 /home/user/PycharmProjects/cda-test-data-generator/src/main.py
```

or

```shell
cd /home/user/PycharmProjects/cda-test-data-generator/src/
python3 main.py
```

both should work

akomii replied:
When running

```shell
python3 main.py --n 10
```

in the src/ folder I get:

```shell
Traceback (most recent call last):
  File "/home/akomi/PycharmProjects/cda-test-data-generator/src/main.py", line 67, in <module>
    main()
  File "/home/akomi/PycharmProjects/cda-test-data-generator/src/main.py", line 60, in main
    csv_to_cda(csv_path, xslt_file)
  File "/home/akomi/PycharmProjects/cda-test-data-generator/src/csv_to_cda.py", line 115, in csv_to_cda
    output_dir = os.environ['output_dir']
                 ~~~~~~~~~~^^^^^^^^^^^^^^
  File "<frozen os>", line 714, in __getitem__
KeyError: 'output_dir'
```

akomii commented on config.toml:
```diff
@@ -0,0 +1,9 @@
+
+[csv_paths]
+cities_csv = "../resources/value_sets/clinics.csv"
+diagnoses_csv = "../resources/value_sets/icd10gm2023.csv"
```
config.toml is a resource, so...

akomii reviewed (changes requested):
Still one technical issues. Also check the usability of the script. It should be easy to use even for non-technical people

Commit:
misc.csv in individual_attributes.csv and requirements.txt fixed

Commit:
Cedis.csv renamend and beschwerden.txt  added in calculate Dependency

Commit:
changed toml to  command

Commit:
changed main so that paths are resolved  relative to the main.py file

Commit:
outputdir groß

Commit:
toml in resources

Commit:
Provide improvements to main.py:
1. Add Logs
2. Add new command line argument (cleanup)
3. Minor refactoring changes

Commit:
Provide improvements to send_xml_async.py
1. Add Logs
2. Add Docstrings
3. Provide better readability and other refactoring changes

Commit:
Refactor map_csv_to_dataframe for better readability and better performance

Commit:
Change city.csv to clinics.csv

Commit:
Refactor tasks. Tuples -> Dictionaries for better readability

akomii commented on src/main.py:
```diff
@@ -1,20 +1,137 @@
+        csv_to_cda(csv_path, xslt_file)
+
+        if cleanup:
+            logging.info("Cleaning up intermediate CSV file...")
```
I would except for cleanup to clean up the raw xmls too

akomii commented on src/csv_to_cda.py:
```diff
@@ -1,54 +1,137 @@
+        tree = add_processing_instructions(etree.ElementTree(xml_root))
+
+        # Save CDA
+        save_xml(tree, os.path.join(cda_path, f'cda_{i}.cda'))
```
The file format is .xml, not .cda

akomii reviewed (changes requested):
Two minor issues. Then its ready to merge

Commit:
cleanup removes Raw

Commit:
.cda -> .xml

akomii committed:
added .idea to .gitignore

akomii committed:
Merge branch 'main' into develop
